File: quwoquan_service/runtime/recommendation/object_relation_edge_type.go
package recommendation

// 对象关系边类型闭集，真相源 contracts/metadata/_shared/types.yaml 的
// ObjectRelationEdgeType，parity 由 quwoquan_ops/gate/verify_object_relation_edge_type_contract.py 守。
//
// 收敛前 edgeType 是无约束 string，词表实际分裂成两个互不相交的半边：物化器只写
// semantic_co_mention / tag_overlap / geo_proximity，端侧 switch 只认
// author_of / posted_to_circle / ...。两边各自消费自己那半，任何一侧都不会报错，
// 于是"对象页关系边"这个共同事实源在生产与消费之间从未真正对上。闭集加上 parity 门
// 让新增类型必须同时落到 metadata、Go 与端侧，drift 变成编译期/门禁期错误。

// ObjectRelationEdgeType 是边表 edgeType 的类型化取值。
type ObjectRelationEdgeType string

// 断言边：由第一方记录直接得出，不需要推断。
const (
	EdgeTypeAuthorOf           ObjectRelationEdgeType = "author_of"
	EdgeTypePostedToCircle     ObjectRelationEdgeType = "posted_to_circle"
	EdgeTypeResharedToCircle   ObjectRelationEdgeType = "reshared_to_circle"
	EdgeTypeMentionsEntity     ObjectRelationEdgeType = "mentions_entity"
	EdgeTypeCommentAboutEntity ObjectRelationEdgeType = "comment_about_entity"
	EdgeTypeCircleUnderEntity  ObjectRelationEdgeType = "circle_under_entity"
	EdgeTypeMemberOf           ObjectRelationEdgeType = "member_of"
	EdgeTypeReviewOf           ObjectRelationEdgeType = "review_of"
)

// 空间边：实体之间的地理关系。
//
// EdgeTypeNear 与 EdgeTypeGeoProximity 不可互换：后者是 conditionProfile.regions
// 相同算出来的共现信号（会随 TTL 退场），前者是断言两个实体在空间上挨着。
const (
	EdgeTypeLocatedIn ObjectRelationEdgeType = "located_in"
	EdgeTypePartOf    ObjectRelationEdgeType = "part_of"
	EdgeTypeNear      ObjectRelationEdgeType = "near"
	EdgeTypeRouteStop ObjectRelationEdgeType = "route_stop"
)

// 计算边：从既有信号派生，带 confidence 与 evidenceRefs，随 TTL 退场。
const (
	EdgeTypeSemanticCoMention    ObjectRelationEdgeType = "semantic_co_mention"
	EdgeTypeTagOverlap           ObjectRelationEdgeType = "tag_overlap"
	EdgeTypeGeoProximity         ObjectRelationEdgeType = "geo_proximity"
	EdgeTypeBehaviorCoEngagement ObjectRelationEdgeType = "behavior_co_engagement"
)

// objectRelationEdgeTypes 保持与 metadata 完全一致的顺序，parity 门按序比对。
var objectRelationEdgeTypes = []ObjectRelationEdgeType{
	EdgeTypeAuthorOf,
	EdgeTypePostedToCircle,
	EdgeTypeResharedToCircle,
	EdgeTypeMentionsEntity,
	EdgeTypeCommentAboutEntity,
	EdgeTypeCircleUnderEntity,
	EdgeTypeMemberOf,
	EdgeTypeReviewOf,
	EdgeTypeLocatedIn,
	EdgeTypePartOf,
	EdgeTypeNear,
	EdgeTypeRouteStop,
	EdgeTypeSemanticCoMention,
	EdgeTypeTagOverlap,
	EdgeTypeGeoProximity,
	EdgeTypeBehaviorCoEngagement,
}

// spatialObjectRelationEdgeTypes 是断言型空间边。它们描述实体间的固有地理关系，
// 不随信号衰减，因此不参与计算边的 TTL 退场。
var spatialObjectRelationEdgeTypes = map[ObjectRelationEdgeType]bool{
	EdgeTypeLocatedIn: true,
	EdgeTypePartOf:    true,
	EdgeTypeNear:      true,
	EdgeTypeRouteStop: true,
}

// ObjectRelationEdgeTypes 返回闭集副本，顺序与 metadata 一致。
func ObjectRelationEdgeTypes() []ObjectRelationEdgeType {
	out := make([]ObjectRelationEdgeType, len(objectRelationEdgeTypes))
	copy(out, objectRelationEdgeTypes)
	return out
}

// ObjectRelationEdgeTypeStrings 返回闭集的 wire 取值，便于构造 BSON 过滤与断言。
func ObjectRelationEdgeTypeStrings() []string {
	out := make([]string, 0, len(objectRelationEdgeTypes))
	for _, value := range objectRelationEdgeTypes {
		out = append(out, string(value))
	}
	return out
}

// ParseObjectRelationEdgeType 把 wire 值收敛为闭集成员；不认识的取值返回 false，
// 由调用方决定丢弃还是报错，禁止静默当作合法类型继续流转。
func ParseObjectRelationEdgeType(raw string) (ObjectRelationEdgeType, bool) {
	candidate := ObjectRelationEdgeType(raw)
	for _, value := range objectRelationEdgeTypes {
		if value == candidate {
			return value, true
		}
	}
	return "", false
}

// IsSpatial 判定是否为断言型空间边。
func (t ObjectRelationEdgeType) IsSpatial() bool {
	return spatialObjectRelationEdgeTypes[t]
}

File: quwoquan_app/lib/core/models/object_relation_edge_type.dart
/// 对象关系边类型闭集。
///
/// 真相源是 `contracts/metadata/_shared/types.yaml` 的 `ObjectRelationEdgeType`，
/// parity 由 `quwoquan_ops/gate/verify_object_relation_edge_type_contract.py` 守。
///
/// 收敛前 `ObjectRelationEdge.edgeType` 是裸 String，词表实际分裂成两个互不相交的
/// 半边：服务侧物化器只写 semantic_co_mention / tag_overlap / geo_proximity，端侧
/// switch 只认 author_of / posted_to_circle / ...，两边各自消费自己那半且都不报错。
/// 收成枚举之后新增类型必须同时落到 metadata、Go 与这里，穷举 switch 会在编译期
/// 逼出未处理分支。
enum ObjectRelationEdgeType {
  authorOf('author_of'),
  postedToCircle('posted_to_circle'),
  resharedToCircle('reshared_to_circle'),
  mentionsEntity('mentions_entity'),
  commentAboutEntity('comment_about_entity'),
  circleUnderEntity('circle_under_entity'),
  memberOf('member_of'),
  reviewOf('review_of'),
  locatedIn('located_in'),
  partOf('part_of'),
  near('near'),
  routeStop('route_stop'),
  semanticCoMention('semantic_co_mention'),
  tagOverlap('tag_overlap'),
  geoProximity('geo_proximity'),
  behaviorCoEngagement('behavior_co_engagement');

  const ObjectRelationEdgeType(this.wire);

  final String wire;

  /// 把 wire 值收敛为枚举；不在闭集内返回 null。
  ///
  /// 返回 null 而不是回落到某个默认类型：未登记的 edgeType 说明服务端词表已经漂移，
  /// 此时静默按别的关系渲染会给用户一句错的关系描述，不如整条边不展示。
  static ObjectRelationEdgeType? tryParse(String? raw) {
    final value = raw?.trim() ?? '';
    if (value.isEmpty) {
      return null;
    }
    for (final candidate in ObjectRelationEdgeType.values) {
      if (candidate.wire == value) {
        return candidate;
      }
    }
    return null;
  }

  /// 断言型空间边：描述实体间固有地理关系，不随信号衰减退场。
  ///
  /// [geoProximity] 不在其列——它是「同域」算出来的共现信号，不等于两个实体真的挨着。
  bool get isSpatial => switch (this) {
    ObjectRelationEdgeType.locatedIn ||
    ObjectRelationEdgeType.partOf ||
    ObjectRelationEdgeType.near ||
    ObjectRelationEdgeType.routeStop => true,
    _ => false,
  };
}
